### ISSUE TITLE
v4: Throttling, Debouncing, and Limiting

### ISSUE THREAD

**Issue by TysonMN** [elmish/Elmish.WPF#271]
`wrapDispatch` was removed from [`v4`](https://github.com/elmish/Elmish.WPF/tree/v4) in PR #256 (because it was deemed inferior to a mutually exclusive feature).  A partial migration path is to implement stream adjustment behavior like throttling, debouncing, and limiting.  This partial migration path will help at least one user (c.f. #269).

As a first step, I want to precisely define what I mean by throttling, debouncing, and limiting, and I want to find existing implementations of these behaviors.

**Comment by Evangelink** [comment]
+1 user for the throttling function. I am using it for a slider.

**Comment by TysonMN** [comment]
A slider is what original motivated me to add `wrapDispatch`.  (See #114.)

**Comment by cmeeren** [comment]
@Evangelink Could you please verify that the throttling is actually necessary? IIRC, when we have discussed this issue previously, it seemed that WPF throttles/debounces itself (e.g. for sliders), and thus it may not be needed at all. AFAIK, so far we haven't had any explanation/demonstration of a use-case where throttling/debouncing/etc. it is actually needed (https://github.com/elmish/Elmish.WPF/issues/269#issuecomment-679956871).

**Comment by Evangelink** [comment]
I might be doing something wrong but I confirm that I had to recently introduce the throttle with the dispatchWrap as described in another ticket because of the slowness when dragging the thumb index on the slider. Actually that's not exactly a slowness but more of a thumb index acting jaggy (doing being jumps forward and backward). I didn't noticed that before but I am not sure if that's a miss, the fact that the model was a lot simpler or because the slider wasn't yet styled. We have changed the control template to make it look like youtube bar so maybe we broke something.

I will try to create a small app with a simple model and my slider to see if that's related. Not sure exactly when I will be able to make this but I will try to have it done by the end of the week.

**Comment by Evangelink** [comment]
@cmeeren I have created a little project which kind of reproduce the behavior (jaggy effect) I am having with the third-party component on my real app https://github.com/Evangelink/Elmish.Wpf.Experiments/tree/main/ElmishThrottle

Launch the app, click and drag the thumb forward and/or backward and you should notice the behavior (frames jump).

**Comment by TysonMN** [comment]
@Evangelink, can you also create and share videos or GIFs of this jaggy behavior and then the same of the improved behavior when throttling?

**Comment by Evangelink** [comment]
Jaggy behavior (there is a bounce back and worth while dragging in only one direction):
![jaggy](https://user-images.githubusercontent.com/11340282/107356632-53b4d080-6ad1-11eb-9591-ccd6ae58b4a9.gif)

With the throttle, there is a bit of latency (thumb is not at the exact mouse position) but the jaggy behavior disappear:
![throttled](https://user-images.githubusercontent.com/11340282/107356279-e2751d80-6ad0-11eb-91d2-e67893746967.gif)

**Comment by cmeeren** [comment]
IMHO both of those are poor UX. Ideally it should be debounced in the UI and the slider should not be locked to the exact value from the Elmish model, to ensure a smooth UI, though I have no idea how to make that work cleanly with MVU (perhaps my one main gripes with it).

**Comment by Evangelink** [comment]
I was digging even more on this subject and noticed that there is a `Delay` property on the binding (I wasn't aware of it) and it does seem to provide an even better UX (I will need to find the right threshold but it looks good):

![2021-02-09_13h26_54](https://user-images.githubusercontent.com/11340282/107363480-80b9b100-6ada-11eb-9278-c4e421097016.gif)

**Comment by TysonMN** [comment]
Can you share a snippet of code in which you use this delay?

**Comment by Evangelink** [comment]
Sure, here you go (note that the important part is on the binding of `Value`):

```xml
<Slider Margin="10" IsSnapToTickEnabled="True"
        IsMoveToPointEnabled="True" IsSelectionRangeEnabled="False" LargeChange="5"
        Maximum="{Binding FrameMaxIndex, FallbackValue=1, TargetNullValue=1}"
        Minimum="0" SmallChange="1"
        Value="{Binding CurrentFrameIndex, Mode=TwoWay, Delay=100}" />
```

You can also see https://github.com/Evangelink/Elmish.Wpf.Experiments/pull/3

**Comment by TysonMN** [comment]
Excellent!

The [BindingBase.Delay documentation](https://docs.microsoft.com/en-us/dotnet/api/system.windows.data.bindingbase.delay) says
> Gets or sets the amount of time, in milliseconds, to wait before updating the binding source after the value on the target changes.
> [...]
> To avoid updating the source object with every keystroke, set the Delay property to a reasonable value to cause the binding to update only after that amount of time has elapsed since the user stopped typing.

In my understanding of the difference between throttling, debouncing, and limiting, this `delay` feature of a WPF binding is implementing debouncing where a single "bounce" is passed along at the end of the relevant interval.  See [this SO answer](https://stackoverflow.com/questions/25991367/difference-between-throttling-and-debouncing-a-function/25991510#25991510), which links to [this beautiful demo](http://demo.nimius.net/debounce_throttle/).  The debouncing in that demo is the same as WPF's binding delay feature (i.e. both pass along a single event at the end...instead of the beginning...of the relevant interval).

Here is a video of that behavior in @Evangelink's after branch after making the highlighted change so that the console logging is visible.
![2021-02-09_09-54-19_513](https://user-images.githubusercontent.com/34664007/107390098-f6af1f80-6abc-11eb-8b13-b13356ca0925.gif)

**Comment by TysonMN** [comment]
> As a first steps, I want to precisely define what I mean by throttling, debouncing, and limiting, and I want to find existing implementations of these behaviors.

Here is [that beautiful demo again of throttling vs debouncing](http://demo.nimius.net/debounce_throttle/). I think each of those also has a choice of passing along an event at the beginning of the corresponding interval or at the end.  Both of those do so at the end.  I think of the difference between throttling and (rate) limiting as whether data is lost.  I think of throttling drops data to ensure a maximum rate is not exceeded.  I think of rate limiting using a queue to store all incoming events (so no data is lost) and then pass them along at a rate that does not exceed a given maximum.  I get the impression that some people define rate limiting the same as I have defined throttling.

**Comment by xperiandri** [comment]
So is dispatch wrapping no longer available at all?
`Delay` on `Binding is not supported on UWP

**Comment by TysonMN** [comment]
> So is dispatch wrapping no longer available at all?

Correct.  (c.f. https://github.com/elmish/Elmish.WPF/issues/364#issuecomment-792398824).

> `Delay` on `Binding is not supported on UWP

Ok.

As I said in https://github.com/elmish/Elmish.WPF/issues/364#issuecomment-792414402, I think we can implement throttling in a way that is compatible with `mapMsg`.  This implementation could exist ing a different NuGet package called `Elmish.WPF.Reactive` that depends on `FSharp.Control.Reactive`.

I am probably oversimplifying there and missing some issues.  However, I don't plan to look into this until after I have the composable binding API complete (c.f. #263).

**Comment by xperiandri** [comment]
@TysonMN so what are we going to do with throttling? So that I can migrate Elmish.Uno to be compatible with Elmish.WPF v4

**Comment by TysonMN** [comment]
What is your use case for throttling?

I haven't spent any time working on throttling, deboucing, and limiting because I am still working on the composable binding API, but I still think they will be possible to implement.

**Comment by xperiandri** [comment]
Entering text to text fields and calling API with that text automatically

**Comment by xperiandri** [comment]
Like autocomplte

**Comment by TysonMN** [comment]
That is a very good use case.  Thanks for mentioning it.

**Comment by TysonMN** [comment]
Great news!

I just added to the `master` branch support composable _monomorphic_ dispatch wrapping.  By monomorphic, I mean the signature of `wrapDispatch` is `(obj -> unit) -> obj -> unit`.  In the `v3` branch, the type of `wrapDispatch` is `('msg -> unit) -> 'msg -> unit`, which is polymorphic (with type parameter `'msg`).

To see this in action, try using the slider in the `SingleCounter` sample in the branch [`example/wrap_dispatch`](https://github.com/TysonMN/Elmish.WPF/tree/example/wrap_dispatch).  It uses [System.Reactive](https://www.nuget.org/packages/System.Reactive).  Only now as I am typing this and rereading this issue did I realize that using [`FSharp.Control.Reactive`](https://www.nuget.org/packages/FSharp.Control.Reactive) would probably be better.  Improving this is the problem for future me.

![2021-08-12_23-02-36_665](https://user-images.githubusercontent.com/34664007/129303237-fd67eb39-3e55-4e1b-9674-6795cc3facc9.gif)

@xperiandri, is this feature sufficient for your use case?

(CC @cmeeren: Just wanted to bring to your attention this great improvement.  No need to replay.)

**Comment by xperiandri** [comment]
What is the main difference to v3 implementation?

**Comment by xperiandri** [comment]
Why signature has changed?

**Comment by TysonMN** [comment]
Sorry for the delayed response.  I haven't had time to write a clear(er) explanation.

And actually, I still don't.  I am going to delay my response further because I have an idea that might allow me to improve this feature.  This might take a week.  Stay tuned!

**Comment by TysonMN** [comment]
> What is the main difference to v3 implementation?
> Why signature has changed?

These answer are a bit moot now (see next comment), but here are brief answers.

The signature of the `wrapDispatch` parameter in v3 is
```fsharp
('msg -> unit) -> 'msg -> unit
```
while the signature of the parameter demonstrated in the above code is
```fsharp
(obj -> unit) -> obj -> unit
```

This feature is unsound in the sense that it is possible for the user to write code involving it that complies but fails at runtime due to an `InvalidCastException`.  However, I think it is better than no dispatch wrapping.

But forget about this.  I now have something better!

**Comment by TysonMN** [comment]
Dispatch wrapping is back!  See the branch [polymorphic_dispatch_wrapping](https://github.com/TysonMN/Elmish.WPF/tree/polymorphic_dispatch_wrapping) for a proof of concept.  (I still need to do lots of clean up.)

![2021-08-19_20-24-34_675](https://user-images.githubusercontent.com/34664007/130165219-b5f21611-0ea3-4a0e-980f-b398c173c0b6.gif)

This time, dispatch wrapping is
- sound,
- polymorphic, and
- covariant in `'msg` (in the sense that the returned dispatch function doesn't have to have the same type as the given dispatch function...and the way in which it is is allowed to differ is consistent with the covariance of `'msg`).

It couldn't be any better!! :D

So @xperiandri, you will not lose any functionality (related to dispatch wrapping) in v4.  In fact, it will be slightly more expressive than it was before.

I opened this issue because I thought I would have to introduce a breaking change in v4 by removing all of the optional `wrapDispatch` parameters.  Now those can be added back so that it is not a breaking change.  I will deprecate them though and remove them in the next major version.  I will create a new issue to track this work.

I don't like the name `wrapDispatch`.  I don't think the user thinks about dispatching (which is essentially enqueuing a message into the Elmish message queue).

Instead, of the name function name `addWrapDispatch`, my current favorite name is `alterMsgStream`.

(CC @cmeeren: Just wanted to bring to your attention this _even greater_ improvement. No need to replay.)

**Comment by xperiandri** [comment]
Sounds amazing! I will have a look

**Comment by cmeeren** [comment]
This is excellent, @TysonMN! 😁 Good work! 👍

**Comment by TysonMN** [comment]
I now better understand the limits of F# against which I was fighting.  Below is a rather minimal reproduction.  It doesn't perfectly reproduce the problem, but it is close.

---------------------------

First, the recursion I need seems particularly difficult for F#'s type inference.  To compensate, I need to make everything explicit: argument types, the return type, and even type parameters.

One of the last things I figure out or tried was specifying the return type.  Without specifying the return type (of the method `Initialize`), I get this error.

> Error FS0006 A use of the function 'Initalize' does not match a type inferred elsewhere. The inferred type of the function is
> `Elmish.WPF.VmBinding2<'model,'msg> -> Microsoft.FSharp.Core.unit -> Elmish.WPF.VmBinding<'model,'msg> Microsoft.FSharp.Core.option`.
> The type of the function required at this point of use is
`Elmish.WPF.VmBinding2<'model,'msg> -> Microsoft.FSharp.Core.unit -> Elmish.WPF.VmBinding<'model,'msg> Microsoft.FSharp.Core.option`
> This error may be due to limitations associated with generic recursion within a 'let rec' collection or within a group of classes. Consider giving a full type signature for the targets of recursive calls including type annotations for both argument and return types.

Notice that the type I provided and the desired type are the same but the F# compiler seems to suggest they are different.  Even so, it was the suggestion at the bottom that lead me to add a type annotation for the return type on each of my recursive functions.

In my reproduction below, there is a less informative _warning_ message when the return type is not specified.
> Warning FS0064 This construct causes code to be less generic than indicated by the type annotations. The type variable 'b has been constrained to be type 'obj'.

I constantly saw this message at other times during my coding with `'model` and `'msg` in place of `'b`.

----------------

Second, in order to have explicit type parameters, I either need to use a method or a function _in a module_.  I cannot use a function in a class.  Trying to do so results in this error.

> Error FS0665 Explicit type parameters may only be used on module or member bindings

Because of this and because so many of the types involved are depend on each other recursively (and because the `'model` and `'msg` type parameters are in scope within `ViewModel<'model, 'msg>`), I think the best way to organize the code is to similar to what I have done in that proof of concept branch.  Specifically, these key recursive functions should be defined as methods in classes other than `ViewModel<_, _>`.

-------------------------

Here is my small (near) reproduction.

```fsharp
module TopLevelModule

[<RequireQualifiedAccess>]
type HeterogeneousValueList<'a> =
  | Empty
  | HeadTail of 'a * HeterogeneousValueList<obj>

[<RequireQualifiedAccess>]
type HeterogeneousFunctionList<'a, 'b> =
  | Empty
  | HeadTail of ('a -> 'b) * HeterogeneousFunctionList<obj, obj>


module HeterogeneousValueList =
  let rec mapAsFunctionBindingInModule<'a, 'b>
      (fs: HeterogeneousFunctionList<'a, 'b>)
      (aList: HeterogeneousValueList<'a>)
      : HeterogeneousValueList<'b> =
    match fs, aList with
    | HeterogeneousFunctionList.Empty, _
    | _, HeterogeneousValueList.Empty ->
        HeterogeneousValueList.Empty
    | HeterogeneousFunctionList.HeadTail(f, fsTail), HeterogeneousValueList.HeadTail (a, aTail) ->
        let b: 'b = f a
        let tail = mapAsFunctionBindingInModule fsTail aTail
        HeterogeneousValueList.HeadTail(b, tail)


type HeterogeneousValueList () =

  //let rec mapAsFunctionBindingInClass<'a, 'b> // FS0665 Explicit type parameters may only be used on module or member bindings
  //    (fs: HeterogeneousFunctionList<'a, 'b>)
  //    (aList: HeterogeneousValueList<'a>)
  //    : HeterogeneousValueList<'b> =
  //  match fs, aList with
  //  | HeterogeneousFunctionList.Empty, _
  //  | _, HeterogeneousValueList.Empty ->
  //      HeterogeneousValueList.Empty
  //  | HeterogeneousFunctionList.HeadTail(f, fsTail), HeterogeneousValueList.HeadTail (a, aTail) ->
  //      let b: 'b = f a
  //      let tail = mapAsFunctionBindingInClass fsTail aTail
  //      HeterogeneousValueList.HeadTail(b, tail)

  static member mapAsMemberBindingInClass<'a, 'b>
      (fs: HeterogeneousFunctionList<'a, 'b>,
       aList: HeterogeneousValueList<'a>)
      : HeterogeneousValueList<'b> =
    match fs, aList with
    | HeterogeneousFunctionList.Empty, _
    | _, HeterogeneousValueList.Empty ->
        HeterogeneousValueList.Empty
    | HeterogeneousFunctionList.HeadTail(f, fsTail), HeterogeneousValueList.HeadTail (a, aTail) ->
        let b: 'b = f a
        let tail = HeterogeneousValueList.mapAsMemberBindingInClass(fsTail, aTail)
        HeterogeneousValueList.HeadTail(b, tail)
```
